### ISSUE TITLE
V6 Proposal Work

### ISSUE THREAD

**Issue by gregglind** [mozilla/shield-studies-addon-utils#278]
## Claim:  It's worth making a simpler, clearer v6 for "Instrumented Extensions"

We now have several years of experience with extensions.

Observations
- extensions will continue to exist for a while
- privileged extension code is a hassle (requires singing, hard to inspect).
- privileged code should be minimized.  Every line of it should rely on things only available through ChromeUtils.import.
- In the Bootstrap Addon days, Cu.import and unload punished having many modules / files.
- In webextension days, `background` handles multiple sequential files just fine.
- **I cannot do engineering support of V5**  It's too hard, and weird, and not self-serve.  WEE's are complex.

## Thus a proposal for V6.

1.  Small, clean webextensionExperiment API providing ONLY

    - 'getInformationFromFirefox'   (cliendId, prefs overrides).  This will be future proof for if normandy ever provides branch information.
    - telemetry sending (recordEvent style only).
    - set extension related prefs

2.  Unprivileged 'study' ES6 class, loaded in `background.js`.    Contains most of the 'standard study' business logic.

3.  Authors subclass `study` as they see fit.

Why this split?
- no one should ever have to inspect or debug the 'privileged' code
- using / setting prefs is useful for devs and QA.  Setting prefs is supported by web-ext-config.js
- All the 'business logic' is surfaced in inspectable code in the extension
- authors can subclass or override any methods they like.  
- automated testing can happen on most of the business logic.
- (switch to recordEvent => get rid of a weird 'shield-study' type.)

Deeper details.
- pref cascade:  study.variation  =>  <addonId>.variation => randomChoice from config
- all parquet events are done instead with recordEvent.

**Comment by gregglind** [comment]
@motin @rhelmer Thoughts are welcome here.

**Comment by jtg567** [comment]
> (switch to recordEvent => get rid of a weird 'shield-study' type.)

I think this is a good direction to go but absolutely must be communicated widely to all consumers of shield data well in advance. I know I have _some_ code that would have to be re-written around this change, but probably not an unmanageable quantity. I cannot speak for anyone else, particularly from data science.

**Comment by motin** [comment]
Update: It is rather clear now that the intention is to not even need a V6 after the corresponding APIs have been moved to tree. 

A quick overview of the most relevant APIs to move to tree are:

```
browser.study.getStudyInfo( )
browser.study.sendTelemetry( payload )
browser.study.searchSentTelemetry( searchTelemetryQuery )
browser.study.searchSeenTelemetry( searchTelemetryQuery )
browser.study.getTestingOverrides( )
browser.study.onReady () Event
browser.study.onEndStudy () Event
browser.study.onBrowserUpdate () Event
browser.study.logger.info( values )
browser.study.logger.log( values )
browser.study.logger.debug( values )
browser.study.logger.warn( values )
browser.study.logger.error( values )
browser.privacyContext.*
```

They will live under various new namespaces, supposedly:

```
browser.normandy.*
browser.telemetry.*
browser.telemetryArchive.*
browser.logger.*
browser.privacyContext.*
```

Detailed planning, including the additional API:s that are necessary for debugging and developing studies are found in https://docs.google.com/spreadsheets/d/1w6zZKCxn68gTDDViO_s7ei8ITIlr1teEUi7rP_0ehAc/edit#gid=0

@rhelmer Do we have a tracking bug already in bugzilla for this work?

**Comment by motin** [comment]
The bug for the Normandy-related changes: https://bugzilla.mozilla.org/show_bug.cgi?id=1522214

**Comment by rhelmer** [comment]
> They will live under various new namespaces, supposedly:
> 
> ```
> browser.normandy.*

This is covered by https://bugzilla.mozilla.org/show_bug.cgi?id=1522214

> ```
> browser.telemetry.*
> browser.telemetryArchive.*
> ```

Do we need both these separate namespaces for telemetry, or could these both go under `browser.telemetry`? Also curious if you know what is missing from https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/webextension-api.html

> `browser.logger.*`

I wonder why `console.[warn,err,log]` aren't good enough here, does this use the more configurable loggers that Firefox provides?

> `browser.privacyContext.*`

Is this for detecting private browsing mode and if so is it equivalent to `chrome.extension.inIncognitoContext` http://developer.chrome.com/dev/extensions/extension.html#property-inIncognitoContext ?

> @rhelmer Do we have a tracking bug already in bugzilla for this work?

@motin I think we do probably want a tracking bug that includes the Normandy one and also those above.

Also one thing that is missing is some way to get and set prefs (and in the future feature gates), I know @biancadanforth has been exploring this. Do you know if there's a reason it wasn't included in the spreadsheet (I see a few mentions of prefs but nothing about a general pref-flipping API, AFAICT)

Thanks!

**Comment by rhelmer** [comment]
> Is this for detecting private browsing mode and if so is it equivalent to `chrome.extension.inIncognitoContext` http://developer.chrome.com/dev/extensions/extension.html#property-inIncognitoContext ?

Firefox does support this too, I should mention: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/inIncognitoContext

**Comment by motin** [comment]
> > ```
> > browser.telemetry.*
> > browser.telemetryArchive.*
> > ```
> 
> Do we need both these separate namespaces for telemetry, or could these both go under `browser.telemetry`? 

They can both just as well go into the same namespace, and/or be nested (eg the exposing of methods of TelemetryArchive.jsm could live under browser.telemetry.archive.*).

> Also curious if you know what is missing from https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/webextension-api.html

I guess the only main thing missing is the ability to encrypt the payloads before sending telemetry in Pioneer experiments, but the encryption can just as well be exposed under `browser.pioneer.*` or `browser.pioneer.*` instead of `browser.telemetry.*`.

> > `browser.logger.*`
> 
> I wonder why `console.[warn,error,log]` aren't good enough here, does this use the more configurable loggers that Firefox provides?

TL;DR: The study-specific messages sent via console.[warn,error,log] can't be controlled by a pref and will thus lead to excessive browser console output in production. Please add that feature to console.[warn,error,log] and we need no browser[.study].logger.*.

Relevant discussion here: https://github.com/mozilla/shield-studies-addon-utils/issues/139#issuecomment-417624890

> > `browser.privacyContext.*`
> 
> > Is this for detecting private browsing mode and if so is it equivalent to `chrome.extension.inIncognitoContext` http://developer.chrome.com/dev/extensions/extension.html#property-inIncognitoContext ?
> 
> Firefox does support this too, I should mention: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/inIncognitoContext

inIncognitoContext is only for content scripts and will thus not work for background scripts. Since background scripts are shared across ordinary and private browsing windows, the purpose is for study add-ons to be able to suspend instrumentation and telemetry submissions while a private browsing session is active. 

`browser.privacyContext.*` currently lives [in the template](https://github.com/mozilla/shield-studies-addon-template/tree/master/src/privileged/privacyContext) only and exposes `PrivateBrowsingUtils.permanentPrivateBrowsing` as well as checks if any private window is open. The latter can probably be achieved by `browser.windows.*` and the former could ideally be added to [`browser.privacy`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/privacy) instead of requiring a separate API.

> > @rhelmer Do we have a tracking bug already in bugzilla for this work?
> 
> @motin I think we do probably want a tracking bug that includes the Normandy one and also those above.

Sounds great, please add! Note that https://bugzilla.mozilla.org/show_bug.cgi?id=1414409 should probably be linked in some way.

> Also one thing that is missing is some way to get and set prefs (and in the future feature gates), I know @biancadanforth has been exploring this. Do you know if there's a reason it wasn't included in the spreadsheet (I see a few mentions of prefs but nothing about a general pref-flipping API, AFAICT)

Yes, we also believed browser.prefs was necessary, but found that it was safer and more study-author-friendly to read and write all necessary prefs internally in the browser.study API implementation. See these entries for more info: https://github.com/mozilla/shield-studies-addon-utils/issues/249#issuecomment-408530099, https://github.com/mozilla/shield-studies-addon-utils/issues/212#issuecomment-417614057

That being said, i can definitely see the value of having a safe version of `browser.prefs` or `browser.featureGates` available for privileged add-ons, since it allows study authors to write less privileged code themselves.

**Comment by motin** [comment]
@rhelmer fyi, I recently closed all irrelevant or fixed issues in this repo and labeled the remaining issues based on what improvement area they relate to (testing, engineering, docs, data review or analysis). Hopefully this helps in improving the understanding of the strengths and shortcomings of the self-contained add-on-based workflows.

**Comment by motin** [comment]
Closing this since relevant feedback and proposals have been iterated and resulted in https://phabricator.services.mozilla.com/D29913
